### PR TITLE
Add structured pruning support for PyTorch models.

### PR DESCRIPTION
**Pull request by lior-dikstein**
This commit introduces a new function 'pytorch_pruning_experimental' to perform structured pruning on PyTorch models

## Checklist before requesting a review:
- [ ] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [ ] All function and files are well documented. 
- [ ] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [ ] The function and variable names are informative. 
- [ ] I have checked for code duplications.
- [ ] I have added new unittest (if necessary).

### CODE REVIEW

File: tests/pytorch_tests/pruning_tests/test_pretrained_models.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                                f"score {max_score_removed}")
+
+        # Validate that the actual compression rate does not exceed the target compression rate.
+        self.assertTrue(actual_cr <= cr,
+                        f"Expected the actual compression rate: {actual_cr} to not exceed the target compression "
+                        f"rate: {cr}")
+
+
+# Function to generate an infinite stream of dummy images and labels
+def dummy_data_generator():
```
Comment (reuvenperetz) [inline tests/pytorch_tests/pruning_tests/test_pretrained_models.py:144]: If you're not using it, then remove it. But I think it's better to add a flag that indicates whether to test it or not, which is disabled by default (and at least test it once when you run it locally to see if the expected post-pruning retraining is possible).

**Comment by reuvenperetz** [comment]
Don't forget to add tutorial and the docs rst.

File: tests/pytorch_tests/pruning_tests/test_pretrained_models.py (hunk 1 above)
Comment (lior-dikstein) [reply tests/pytorch_tests/pruning_tests/test_pretrained_models.py:144]: I added retraining

File: tutorials/notebooks/pytorch/pruning/example_pytorch_pruning_mnist.ipynb [hunk 2]
```diff
@@ -0,0 +1,412 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "source": [
+    "# Structured Pruning of a Fully-Connected PyTorch Model\n",
+    "\n",
+    "Welcome to this tutorial, where we will guide you through the process of training, pruning, and retraining a fully connected neural network model using the PyTorch framework. The tutorial is organized in the following sections: \n",
```
Comment (reuvenperetz) [inline tutorials/notebooks/pytorch/pruning/example_pytorch_pruning_mnist.ipynb:8]: Please add a google colab link for this tutorial.

File: tutorials/notebooks/pytorch/pruning/example_pytorch_pruning_mnist.ipynb [hunk 3]
```diff
@@ -0,0 +1,412 @@
+{
+ "cells": [
+  {
```
Comment (reuvenperetz) [inline tutorials/notebooks/pytorch/pruning/example_pytorch_pruning_mnist.ipynb:3]: Add a reference to this tutorial in the notebooks README.

**Comment by reuvenperetz** [comment]
Please do not forget to update MCT README.

File: tutorials/notebooks/pytorch/pruning/example_pytorch_pruning_mnist.ipynb (hunk 2 above)
Comment (lior-dikstein) [reply tutorials/notebooks/pytorch/pruning/example_pytorch_pruning_mnist.ipynb:8]: I need the file to be pushed first or else the test fails on missing link

File: tutorials/notebooks/pytorch/pruning/example_pytorch_pruning_mnist.ipynb [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+    "collapsed": false
+   },
+   "id": "a3eaaa9bb34ebf71"
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "outputs": [],
+   "source": [
+    "pruned_model_retrained = train_model(pruned_model, train_loader, test_loader, device, epochs=1)"
```
[the remark this answers is not in the document]
Comment (lior-dikstein) [reply tutorials/notebooks/pytorch/pruning/example_pytorch_pruning_mnist.ipynb:350]: yes I will fix it

**Comment by reuvenperetz** [comment]
Please update the notebooks readme https://github.com/sony/model_optimization/blob/main/tutorials/notebooks/README.md

**Review by reuvenperetz — APPROVED** [review]
